keps/sig-node/4815-dra-partitionable-devices/README.md

```diff
@@ -0,0 +1,2093 @@
+  partitions the same. In 1.31, it is already possible to select a device
+  partition via a `Resourceclaim`, so long as the device vendor has
+  pre-partitioned its devices and advertised them in a `ResourceSlice`. The
+  user-facing mechanism to select these devices shouldn't change -- only the
+  ability for the vendor to advertise "overlapping" partitions, such that the
+  scheduler will never allocate conflicting ones at the same time.
+
+* Provide abstractions to keep the overall footprint of a ResourceSlice small,
+  such that all full devices (and their partitions) can be represented in a
+  concise way.
+
+### Non-Goals
+
+* Allow a user to allocate arbitrarily-sized partitions of a device. In other
+  words, only those partitions explicitly defined by a ResourceSlice will be
+  allocatable by a user. Arbitrary slicing is out of scope for this KEP.
+
+## Proposal
+
+The basic idea is the following:
+
+1. Introduce a new device type called `CompositeDevice` which has the same
+   fields as a `BasicDevice`, plus two more. The first is a field called
+   `Includes` and the second is a field called `ConsumesCapacityFrom`. Both
+   full devices and their partitions are represented as instances of this new
+   `CompositeDevice` type and are listed right next to one another in the
+   top-level `Devices` list of a `ResourceSlice`.
+
+1. The `Includes` field serves to reference a set of "mixins" that a
+   `CompositeDevice` can reference to extend the set of attributes and
+   capacities it defines explicitly. The goal being to reduce the overall
+   footprint of each device when a common set of attributes and capacities can
+   be applied. The mixins themselves are introduced as a list of top-level
+   objects directly next to the list of `Devices` inside a `ResourceSlice`.
+   They are not allocatable on their own.
+
+1. The `ConsumesCapacityFrom` field contains a list of *other* devices where the
+   capacity of the current device should be consumed if the scheduler decides
+   to allocate it. This essentially removes that capacity from any referenced
+   devices, rendering them unallocatable on their own.
+
+With these additions in place, the scheduler has everything it needs to support
+the dynamic allocation of both full devices and their (possibly overlapping)
+fixed-size partitions. That is to say, the scheduler now has the ability to
+"flatten" all devices by applying any mixins from their `Includes` fields as
+well as track any capacities consumed from one device by another through its
+`ConsumesCapacityFrom` field. More details on the actual algorithm the
+scheduler follows to make allocation decisions based on the
+`ConsumesCapacityFrom` field can be found in the Design Details section below.
+
+## Design Details
+
+The exact set of proposed API changes can be seen below:
+```go
+// ResourceSliceSpec contains the information published by the driver in one ResourceSlice.
+type ResourceSliceSpec struct {
```

Comment:
ResourceSlices are defined per node (roughly), so there will be a lot of repeated YAMLs. Did you consider having the definition of all possible composite devices somewhere else? Like the DeviceClass?
Or does it have to be that dynamic?

cc @johnbelamaric @thockin

Reply:
Yes, we discussed this quite a bit. See the following for a diccussion on all the options we considered:
https://github.com/kubernetes-sigs/wg-device-management/issues/20#issuecomment-2168189769

In particular, this comment related to Option 7, which is most relevant to your question: 
https://github.com/kubernetes-sigs/wg-device-management/pull/35#issuecomment-2173627898

Reply:
Please add the conclusions of those conversations to the Alternatives section.

Reply:
Done